### PR TITLE
endpoint development

### PR DESCRIPTION
**Pull request by vatsal278**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -20,7 +20,8 @@ require (
 	github.com/go-playground/validator/v10 v10.11.0 // indirect
 	github.com/google/go-cmp v0.5.9 // indirect
 	github.com/leodido/go-urn v1.2.1 // indirect
+	github.com/vatsal278/msgbroker v1.0.3 // indirect
 	github.com/vatsal278/user-mgmt-svc v0.0.0-20221019051737-6bdc60a4d152 // indirect
```
Comment (PereRohit) [inline go.mod]: This should not be imported. 
`go mod tidy` should remove else find where this is used and remove.

File: internal/config/config.go [hunk 2]
```diff
@@ -87,11 +97,25 @@ func InitSvcConfig(cfg Config) *SvcConfig {
 	// init required services and assign to the service struct fields
 	dataBase := Connect(cfg.DataBase, cfg.DataBase.TableName)
 	jwtSvc := jwtSvc.JWTAuthService(cfg.SecretKey)
+	msgBrokerSvc := sdk.NewMsgBrokerSvc(cfg.MessageQueue.SvcUrl)
+	err := msgBrokerSvc.RegisterSub("PUT", "http://localhost/activate", "", cfg.MessageQueue.ActivatedAccountChannel)
```
Comment (PereRohit) [inline internal/config/config.go]: activation flow is pending. Remove this logic.

File: internal/handler/handler.go [hunk 3]
```diff
@@ -42,3 +52,44 @@ func (svc userMgmtSvc) HealthCheck() (svcName string, msg string, stat bool) {
 	set = true
 	return
 }
+func (svc userMgmtSvc) SignUp(w http.ResponseWriter, r *http.Request) {
+	var credential model.SignUpCredentials
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Error(err)
+		response.ToJson(w, http.StatusBadRequest, "unable to read request body", nil)
```
Comment (PereRohit) [inline internal/handler/handler.go]: Can this be modified to send the messages with error codes? This is applicable for all the responses.

File: internal/logic/logic.go [hunk 4]
```diff
@@ -1,34 +1,185 @@
...
 	}
 }
 
 func (l userMgmtSvcLogic) HealthCheck() bool {
 	// check all internal services are working fine
 	return l.DsSvc.HealthCheck()
 }
+
+func (l userMgmtSvcLogic) Signup(credential model.SignUpCredentials) *respModel.Response {
+	if credential.Email == "" || credential.Password == "" || credential.Name == "" {
```
Comment (PereRohit) [inline internal/logic/logic.go]: Use `github.com/PereRohit/util/validator` to validate the struct fields by adding the `validate` struct tags. This should validate the required and specific format fields without the need for every field validation manually.

This validation should be done at the handler level.

File: internal/logic/logic.go [hunk 5]
```diff
@@ -1,34 +1,185 @@
...
+			Message: "Provide required details",
+			Data:    nil,
+		}
+	}
+	result, err := l.DsSvc.Get(map[string]interface{}{"email": credential.Email})
+	if err != nil {
+		log.Error(err)
+		return &respModel.Response{
+			Status:  http.StatusInternalServerError,
+			Message: "Problem creating account",
```
Comment (PereRohit) [inline internal/logic/logic.go]: Use error codes for failure or any other error cases that needs to be sent back to the client. Applicable for all.

File: internal/repo/bCrypt/encrypt.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	"log"
+)
+
+// GeneratePasswordHash handles generating password hash
+// bCrypt hashes password of type byte
+func GeneratePasswordHash(password []byte) string {
+	// default cost is 10
+	hashedPassword, err := bcrypt.GenerateFromPassword(password, bcrypt.DefaultCost)
+	if err != nil {
+		log.Fatal(err)
```
Comment (PereRohit) [inline internal/repo/bCrypt/encrypt.go]: Return error instead of fatal here. This can crash the server.

File: internal/logic/logic.go [hunk 7]
```diff
@@ -1,34 +1,185 @@
...
+
+	newUser := model.User{
+		Id:           uuid.New().String(),
+		Email:        credential.Email,
+		Password:     credential.Password,
+		Name:         credential.Name,
+		Company:      "perennial",
+		RegisteredOn: credential.RegistrationTimestamp,
+	}
+	salt := bCrypt.GeneratePasswordHash([]byte(newUser.Password))
```
Comment (PereRohit) [inline internal/logic/logic.go]: The password needs to be stored in encrypted format. The encryption is to be done after adding the salt to the given password and then performing encryption.

File: internal/logic/logic.go [hunk 8]
```diff
@@ -1,34 +1,185 @@
...
+	if err != nil {
+		log.Error(err.Error())
+		return &respModel.Response{
+			Status:  http.StatusBadRequest,
+			Message: "Problem creating an account",
+			Data:    nil,
+		}
+	}
+	userID := fmt.Sprintf(`{"user_id":"%s"}`, newUser.Id)
+	err = l.msgQueue.MsgBroker.PushMsg(userID, l.msgQueue.PubId, l.msgQueue.Channel)
```
Comment (PereRohit) [inline internal/logic/logic.go]: Put this block in a goroutine and return immediately once the insertion is successful.

File: internal/logic/logic.go [hunk 9]
```diff
@@ -1,34 +1,185 @@
...
+		return &respModel.Response{
+			Status:  http.StatusUnauthorized,
+			Message: "User account was not found",
+			Data:    nil,
+		}
+	}
+	if result[0].Active != true {
+		return &respModel.Response{
+			Status:  http.StatusAccepted,
+			Message: "Problem logging into your account",
```
Comment (PereRohit) [inline internal/logic/logic.go]: The data field should be empty and the message should be `Account activation in progress`

File: internal/logic/logic.go [hunk 10]
```diff
@@ -1,34 +1,185 @@
...
+		return &respModel.Response{
+			Status:  http.StatusInternalServerError,
+			Message: "Problem notifying account management service",
+			Data:    nil,
+		}
+	}
+
+	return &respModel.Response{
+		Status:  http.StatusCreated,
+		Message: "SUCCESS",
```
Comment (PereRohit) [inline internal/logic/logic.go]: Check the spec and confirm what is to be sent in the response.

File: internal/logic/logic.go [hunk 11]
```diff
@@ -1,34 +1,185 @@
...
+	id := result[0].Id
+	jwtToken, err := l.jwtService.GenerateToken(jwt.SigningMethodHS256, id, 6)
+	if err != nil {
+		return &respModel.Response{
+			Status:  http.StatusInternalServerError,
+			Message: "Unable to generate jwt token",
+			Data:    nil,
+		}
+	}
+	http.SetCookie(w, &http.Cookie{
```
Comment (PereRohit) [inline internal/logic/logic.go:176]: There are other parameters as well that needs to be set. Check the Cookie struct and see what the fields represent. Set them with appropriate values.

File: internal/logic/logic.go [hunk 12]
```diff
@@ -1,34 +1,185 @@
...
+		}
+	}
+	if result[0].Email == "" {
+		return &respModel.Response{
+			Status:  http.StatusUnauthorized,
+			Message: "User account was not found",
+			Data:    nil,
+		}
+	}
+	if result[0].Active != true {
```
Comment (PereRohit) [inline internal/logic/logic.go:153]: Check for activation once the email & password matches.

File: internal/logic/logic.go [hunk 13]
```diff
@@ -1,34 +1,185 @@
...
+		return &respModel.Response{
+			Status:  http.StatusAccepted,
+			Message: "Problem logging into your account",
+			Data:    "Account activation in progress",
+		}
+	}
+
+	hashedPassword := []byte(result[0].Salt)
```
Comment (PereRohit) [inline internal/logic/logic.go]: Combine the salt for the email(obtained from DB) with the password in the request.
Pass this new password through the encryption flow to get the encrypted password.
Get the record for this user with this email and new generated password from the DB and return appropriate error if no records are found.

File: internal/model/request.go [hunk 14]
```diff
@@ -1,12 +1,16 @@
...
+import "time"
+
 type PingRequest struct {
 	Data string `json:"data" validate:"required"`
 }
 type SignUpCredentials struct {
-	Name     string `json:"name" binding:"required"`
-	Email    string `json:"email" binding:"required"`
-	Password string `json:"password" binding:"required"`
+	Name                  string    `json:"name" binding:"required"`
```
Comment (PereRohit) [inline internal/model/request.go]: Don't think binding tag is used. Instead use the `validate` tag as per this

> Use github.com/PereRohit/util/validator to validate the struct fields by adding the validate struct tags. This should validate the required and specific format fields without the need for every field validation manually.
> 
> This validation should be done at the handler level.

File: internal/logic/logic.go [hunk 15]
```diff
@@ -1,34 +1,185 @@
...
+	"github.com/PereRohit/util/log"
+	respModel "github.com/PereRohit/util/model"
+	"github.com/dgrijalva/jwt-go"
+	"github.com/google/uuid"
+	"github.com/vatsal278/UserManagementService/internal/config"
+	"github.com/vatsal278/UserManagementService/internal/model"
+	jwtSvc "github.com/vatsal278/UserManagementService/internal/repo/authentication"
+	"github.com/vatsal278/UserManagementService/internal/repo/bCrypt"
 	"github.com/vatsal278/UserManagementService/internal/repo/datasource"
+	"github.com/vatsal278/UserManagementService/internal/repo/helpers"
```
Comment (PereRohit) [inline internal/logic/logic.go]: Where is this used?

File: internal/repo/helpers/loginService.go [hunk 16]
```diff
@@ -0,0 +1,21 @@
+package helpers
```
Comment (PereRohit) [inline internal/repo/helpers/loginService.go]: Don't see the use for this package. Remove if not required.

File: model/datasource.go [hunk 17]
```diff
@@ -1,49 +0,0 @@
-package model
```
Comment (PereRohit) [inline model/datasource.go:1]: Why is this removed?

**Review by PereRohit — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (vatsal278) [reply go.mod]: done

File: internal/config/config.go (hunk 2 above)
Comment (vatsal278) [reply internal/config/config.go]: done

File: internal/logic/logic.go (hunk 4 above)
Comment (vatsal278) [reply internal/logic/logic.go]: done

File: internal/repo/bCrypt/encrypt.go (hunk 6 above)
Comment (vatsal278) [reply internal/repo/bCrypt/encrypt.go]: done

File: internal/logic/logic.go (hunk 7 above)
Comment (vatsal278) [reply internal/logic/logic.go]: done //doubt

File: internal/handler/handler.go (hunk 3 above)
Comment (vatsal278) [reply internal/handler/handler.go]: done

File: internal/logic/logic.go (hunk 5 above)
Comment (vatsal278) [reply internal/logic/logic.go]: done

File: internal/logic/logic.go (hunk 8 above)
Comment (vatsal278) [reply internal/logic/logic.go]: done

File: internal/logic/logic.go (hunk 9 above)
Comment (vatsal278) [reply internal/logic/logic.go]: in readme it says data field should be Account activation in progress

File: internal/logic/logic.go (hunk 10 above)
Comment (vatsal278) [reply internal/logic/logic.go]: done

File: internal/logic/logic.go (hunk 11 above)
Comment (vatsal278) [reply internal/logic/logic.go:176]: other than jwt token we need to set the user id as part of cookie but it is already inside the jwt token which we generated by adding user id to token claims

File: internal/logic/logic.go (hunk 12 above)
Comment (vatsal278) [reply internal/logic/logic.go:153]: done

File: internal/logic/logic.go (hunk 13 above)
Comment (vatsal278) [reply internal/logic/logic.go]: done

File: internal/model/request.go (hunk 14 above)
Comment (vatsal278) [reply internal/model/request.go]: done

File: internal/logic/logic.go (hunk 15 above)
Comment (vatsal278) [reply internal/logic/logic.go]: removed 
done

File: internal/repo/helpers/loginService.go (hunk 16 above)
Comment (vatsal278) [reply internal/repo/helpers/loginService.go]: done

File: internal/logic/logic.go [hunk 18]
```diff
@@ -1,34 +1,201 @@
...
+			Status:  http.StatusBadRequest,
+			Message: codes.GetErr(codes.ErrEmailExists),
+			Data:    nil,
+		}
+	}
+	newUser := model.User{
+		Id:           uuid.New().String(),
+		Email:        credential.Email,
+		Name:         credential.Name,
+		Company:      "perennial",
```
Comment (PereRohit) [inline internal/logic/logic.go:72]: Keep the company name as empty. This will be added later when the user edits their details.

File: internal/logic/logic.go [hunk 19]
```diff
@@ -1,34 +1,201 @@
...
+		if err != nil {
+			log.Error(err)
+			return
+		}
+		return
+	}()
+	return &respModel.Response{
+		Status:  http.StatusCreated,
+		Message: codes.GetErr(codes.Success),
+		Data:    codes.GetErr(codes.AccActivationInProcess),
```
Comment (PereRohit) [inline internal/logic/logic.go:108]: This is not an error. Just return the text as "Activation in progress" in the data field.

File: internal/logic/logic.go [hunk 20]
```diff
@@ -1,34 +1,201 @@
...
+	}
+	if len(result) == 0 {
+		return &respModel.Response{
+			Status:  http.StatusUnauthorized,
+			Message: codes.GetErr(codes.AccNotFound),
+			Data:    nil,
+		}
+	}
+	hashedPassword, err := crypto.GeneratePasswordHash([]byte(credential.Password), []byte(result[0].Id))
+	if err != nil {
```
Comment (PereRohit) [inline internal/logic/logic.go:130]: log the error.

File: internal/logic/logic.go [hunk 21]
```diff
@@ -1,34 +1,201 @@
...
+		log.Error(err.Error())
+		return &respModel.Response{
+			Status:  http.StatusInternalServerError,
+			Message: codes.GetErr(codes.ErrCreatingAccount),
+			Data:    nil,
+		}
+	}
+	go func() {
+		userID := fmt.Sprintf(`{"user_id":"%s"}`, newUser.Id)
+		err = l.msgQueue.MsgBroker.PushMsg(userID, l.msgQueue.PubId, l.msgQueue.Channel)
```
Comment (PereRohit) [inline internal/logic/logic.go:98]: This will cause race condition. Create a new variable in this functions scope using `:=`.
Basically, update `err =` to `err := ` and pass the required field(s) as arguments to this function.

File: internal/repo/crypto/encrypt.go [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+
+import (
+	"crypto/sha512"
+	"encoding/hex"
+)
+
+func GeneratePasswordHash(password []byte, salt []byte) (string, error) {
+	password = append(password, salt...)
+	var sha512Hasher = sha512.New()
+	sha512Hasher.Write(password)
```
Comment (PereRohit) [inline internal/repo/crypto/encrypt.go:11]: Handle the error.

**Review by PereRohit — APPROVED** [review]
Minor comments. Please update PR and proceed to merge.

File: internal/logic/logic.go (hunk 21 above)
Comment (vatsal278) [reply internal/logic/logic.go:98]: done

File: internal/logic/logic.go (hunk 20 above)
Comment (vatsal278) [reply internal/logic/logic.go:130]: done

File: internal/logic/logic.go (hunk 19 above)
Comment (vatsal278) [reply internal/logic/logic.go:108]: done

File: internal/logic/logic.go (hunk 18 above)
Comment (vatsal278) [reply internal/logic/logic.go:72]: done

File: internal/repo/crypto/encrypt.go (hunk 22 above)
Comment (vatsal278) [reply internal/repo/crypto/encrypt.go:11]: done
